cpp/src/dictionary/add_keys.cu

```diff
@@ -76,33 +77,35 @@ std::unique_ptr<column> add_keys(
     stream);
   // now create the indices column -- map old values to the new ones
   // gather([4,0,3,1,2,2,2,4,0],[0,1,2,3,5]) = [5,0,3,1,2,2,2,5,0]
-  column_view indices_view(data_type{type_id::INT32},
+  column_view indices_view(dictionary_column.indices().type(),
                            dictionary_column.size(),
-                           dictionary_column.indices().data<int32_t>(),
+                           dictionary_column.indices().head(),
                            nullptr,
                            0,
                            dictionary_column.offset());
+  // the result may contain nulls if the input contains nulls
+  // and the corresponding index is therefore invalid/undefined
   auto table_indices = cudf::detail::gather(table_view{{map_indices->view()}},
                                             indices_view,
                                             cudf::detail::out_of_bounds_policy::IGNORE,
                                             cudf::detail::negative_index_policy::NOT_ALLOWED,
                                             mr,
                                             stream)
                          ->release();
-  // the result may contain nulls if the input contains nulls and the corresponding index is
-  // therefore invalid
-  auto contents       = table_indices.front()->release();
-  auto indices_column = std::make_unique<column>(data_type{type_id::INT32},
-                                                 dictionary_column.size(),
-                                                 std::move(*(contents.data.release())),
-                                                 rmm::device_buffer{0, stream, mr},
-                                                 0);
+  // The output of lower_bound is INT32 so we need to convert to UINT32
+  auto const gather_result = table_indices.front()->view();
+  auto indices_column =
```

Comment:
Can this use the relatively new function `cudf::logical_cast`?  (defined in `column_view.hpp`)

```suggestion
  auto indices_column_view = cudf::logical_cast(gather_result, data_type{type_id::UINT32});
```

Reply:
I need a real `column` and not `column_view` to create the dictionary. This does it in one step. Also, this is just a place-holder since the actual indices type may be a different width in the next PR which allows dictionary indices of different widths.
I suppose we could cobble a `column` from the `lower_bound` output data without creating a 2nd column if the widths are the same (as they are in this case). I will try that.